asapdiscovery-data/asapdiscovery/data/utils.py

```diff
@@ -159,6 +164,64 @@ def edit_pdb_file(
     print(f"Wrote {pdb_out}", flush=True)
 
 
+def extract_compounds_from_filenames(fn_list, xtal_pat, compound_pat, fail_val=None):
+    """
+    Extract a list of (xtal, compound_id) from fn_list.
+
+    Parameters
+    ----------
+    fn_list : List[str]
+        List of filenames
+    xtal_pat : Union[str, function]
+        Regex pattern or function for extracting crystal structure ID from filename. If
+        a function is passed, it is expected to return a single str giving the xtal name
+    compound_pat : Union[str, function]
+        Regex pattern or function for extracting crystal structure ID from filename. If
+        a function is passed, it is expected to return a single str giving the
+        compound_id
+    fail_val : str, optional
+        If a value is passed, this value will be returned from the re searches if a
+        match isn't found. If None (default), a ValueError will be raised from the re
+        search
+
+    Returns
+    -------
+    List[Tuple[str, str]]
+        List of (xtal, compound_id)
+    """
+    import re
+
+    if callable(xtal_pat):
+        # Just use the passed function
+        xtal_func = xtal_pat
+    else:
+        # Construct a function for re searching
+        def xtal_func(fn):
```

Comment:
Could seperate this into a function itself, `construct_regex_function` but up to you.

Reply:
done in [1366284](https://github.com/choderalab/asapdiscovery/pull/198/commits/13662848623d9d9057b1d06bcfd862bbec37cd1a)